js/basic-cad-text-electron.js

```diff
@@ -0,0 +1,231 @@
+
+
+  // console.log(textasSVG)
+  // setTimeout(function() {
+  //   var svg = "<svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 50 50\" width=\"150px\"> <path fill=\"#F7931E\" stroke=\"#000\" d=\"" + textasSVG._result + "\"/>  </svg>"
+  //   // console.log(svg)
+  //   return lwsvgparser.loadFromString(svg).then(function(element) {
+  //       return lwsvgparser.parse().then(function(tags) {
+  //         lwsvgparser.editor = {
+  //           name: "Opentype.js",
+  //           version: "1.00"
+  //         };
+  //         drawFile("Text: " + string + " (" + font[0] + ")", tags, true);
+  //         resetView();
+  //       });
+  //     })
+  //     .catch(function(error) {
+  //       console.error('error:', error);
+  //       // $("#addShapeText").modal("hide");
+  //       resetView();
+  //     });
+  //
+  //   printLog('SVG Opened', msgcolor, "file");
+  // }, 2000);
+}
+
+
+// FONTS
+
+// Fetch GoogleFonts List
+// $.get('https://www.googleapis.com/webfonts/v1/webfonts?key=AIzaSyDOcn3KpnYV-8SDuILln1YAH3FWT4K8G38', function (result) {
+//   console.log(result)
+// });
+
+// from https://github.com/nraynaud/webgcode/blob/66e2662fcb72219024976610a6c66d307af84882/webapp/cnc/cam/text.js
+var getFont = function(url) {
+  if (url.match('^http://')) {
+    url = url.replace("http://", "//")
+  }
+  if (url.match('^https://')) {
+    url = url.replace("https://", "//")
+  }
+  return new RSVP.Promise(function(resolve, reject) {
+    opentype.load(url, function(err, font) {
+      if (err)
+        reject();
+      else
+        resolve(font);
+    });
+  });
+};
+
+async function getFontList() {
+  return await fetch('fonts/fontslib/fontslib.json')
+    .then(response => response.json())
+    .then(data => data.items);
+}
+
+function getTextFromData(fontData, fontVariant, text, fontSize, x, y) {
+  if (fontVariant == null)
+    fontVariant = 'regular';
+  x = x == null ? 0 : x;
+  y = y == null ? 0 : y;
+  return getTextFromFile(fontData.files[fontVariant], text, fontSize, x, y);
+}
+
+function getTextFromFile(file, text, fontSize, offsetX, offsetY) {
+  return getFont(file).then(function(font) {
+    console.log(font)
+    var path = font.getPath(text, 0, 0, fontSize);
+    var res = '';
+
+    function xy(x, y) {
+      return (offsetX + x) + ',' + (offsetY - y);
+    }
+
+    for (var i = 0; i < path.commands.length; i++) {
+      var c = path.commands[i];
+      res += ' ' + c.type;
+      if (c.type == 'M' || c.type == 'L')
+        res += ' ' + xy(c.x, c.y);
+      else if (c.type == 'Q')
+        res += xy(c.x1, c.y1) + ' ' + xy(c.x, c.y);
+      else if (c.type == 'C')
+        res += xy(c.x1, c.y1) + ' ' + xy(c.x2, c.y2) + ' ' + xy(c.x, c.y);
+    }
+    // console.log(res)
+    return res;
+  })
+}
+
+function searchFontInList(fontList, fontFamily) {
+  for (var i = 0; i < fontList.length; i++) {
+    var font = fontList[i];
+    if (font.family == fontFamily)
+      return font;
+  }
+  throw {
+    name: 'FontNotFound'
```

Comment:
What is this doing? I'm not familiar with this approach for throwing an exception.

Reply:
I'm not actually sure.  This is a copy of the original basic-cad-text.js file.  My main change was on lines 88-92.

Reply:
ok, you can leave it as is. I thought it was new code.